File: lection04/app_java/src/main/java/ru/hse/lection04/presentationlayer/adapter/LogAdapter.java
package ru.hse.lection04.presentationlayer.adapter;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.ListAdapter;

import ru.hse.lection04.R;
import ru.hse.lection04.objects.LogEntry;

public class LogAdapter extends ListAdapter<LogEntry, LogViewHolder> {
    public LogAdapter() {
        super(new Differ());
    }


    @NonNull
    @Override
    public LogViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        final LayoutInflater inflater = LayoutInflater.from(parent.getContext());
        final View view = inflater.inflate(R.layout.item_log, parent, false);

        return new LogViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull LogViewHolder holder, int position) {
        final LogEntry item = getItem(position);
        holder.bind(item);
    }


    protected static class Differ extends DiffUtil.ItemCallback<LogEntry> {
        @Override
        public boolean areItemsTheSame(@NonNull LogEntry oldItem, @NonNull LogEntry newItem) {
            return oldItem.time.equals(newItem.time);
        }

        @Override
        public boolean areContentsTheSame(@NonNull LogEntry oldItem, @NonNull LogEntry newItem) {
            return oldItem.message.equals(newItem.message);
        }
    }
}

File: lection04/app_kotlin/src/main/java/ru/hse/lection04/presentationlayer/adapter/LogAdapter.kt
package ru.hse.lection04.presentationlayer.adapter

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import ru.hse.lection04.R
import ru.hse.lection04.objects.LogEntry

class LogAdapter : ListAdapter<LogEntry?, LogViewHolder?>(Differ()) {
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): LogViewHolder {
        val inflater = LayoutInflater.from(parent.context)
        val view = inflater.inflate(R.layout.item_log, parent, false)
        return LogViewHolder(view)
    }

    override fun onBindViewHolder(holder: LogViewHolder, position: Int) {
        val item = getItem(position)
        holder.bind(item)
    }

    protected class Differ : DiffUtil.ItemCallback<LogEntry?>() {
        override fun areItemsTheSame(oldItem: LogEntry, newItem: LogEntry): Boolean {
            return oldItem.time == newItem.time
        }

        override fun areContentsTheSame(oldItem: LogEntry, newItem: LogEntry): Boolean {
            return oldItem.message == newItem.message
        }
    }
}